pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/NoDictionaryOnHeapDictionaryJointRule.java

```diff
@@ -0,0 +1,245 @@
+import org.apache.pinot.core.query.request.context.utils.BrokerRequestToQueryContextConverter;
+import org.apache.pinot.core.requesthandler.BrokerRequestOptimizer;
+import org.apache.pinot.core.requesthandler.PinotQueryParserFactory;
+import org.apache.pinot.parsers.AbstractCompiler;
+import org.apache.pinot.sql.parsers.SqlCompilationException;
+import org.apache.pinot.controller.recommender.io.ConfigManager;
+import org.apache.pinot.controller.recommender.io.InputManager;
+import org.apache.pinot.controller.recommender.rules.AbstractRule;
+import org.apache.pinot.controller.recommender.rules.io.params.NoDictionaryOnHeapDictionaryJointRuleParams;
+import org.apache.pinot.controller.recommender.rules.utils.FixedLenBitset;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static org.apache.pinot.controller.recommender.rules.io.params.RecommenderConstants.NoDictionaryOnHeapDictionaryJointRule.*;
+import static org.apache.pinot.controller.recommender.rules.io.params.RecommenderConstants.REALTIME;
+
+
+public class NoDictionaryOnHeapDictionaryJointRule extends AbstractRule {
+  private final Logger LOGGER = LoggerFactory.getLogger(NoDictionaryOnHeapDictionaryJointRule.class);
+  private final BrokerRequestOptimizer _brokerRequestOptimizer = new BrokerRequestOptimizer();
+  private final NoDictionaryOnHeapDictionaryJointRuleParams _params;
+
+  public NoDictionaryOnHeapDictionaryJointRule(InputManager inputManager, ConfigManager outputManager) {
+    super(inputManager, outputManager);
+    _params = inputManager.getNoDictionaryOnHeapDictionaryJointRuleParams();
+  }
+
+  @Override
+  public void run() {
+    LOGGER.info("Recommending no dictionary and on-heap dictionaries");
+
+    int numCols = _inputManager.getNumCols();
+    double[] filterGroupByWeights = new double[numCols];
+    double[] selectionWeights = new double[numCols];
+    AtomicDouble totalWeight = new AtomicDouble(0);
+
+    //**********No dictionary recommendation*******/
+    Set<String> noDictCols = new HashSet<>(_inputManager.getColNameToIntMap().keySet());
+
+    //Exclude cols with index
+    noDictCols.removeAll(_outputManager.getIndexConfig().getInvertedIndexColumns());
+    noDictCols.removeAll(_outputManager.getIndexConfig().getSortedColumn());
+    // TODO: Remove this after range index is implemented for no-dictionary
+    noDictCols.removeAll(_outputManager.getIndexConfig().getRangeIndexColumns());
+    LOGGER.debug("noDictCols {}", noDictCols);
+    //Find out columns used in filter&groupby and selection and corresponding frequencies
+    _inputManager.getQueryWeightMap().forEach((query, weight) -> {
+      parseQuery(query, weight, filterGroupByWeights, selectionWeights);
+      totalWeight.addAndGet(weight);
+    });
+
+    //Add dictionary on columns used in filter&groupby , with frequency > threshold
+    for (int i = 0; i < numCols; i++) {
```

Comment:
Not sure if I follow this. If the column is in filter and group by, why do we have to consider the frequency?

Reply:
Add a comment explaining the rationale

Reply:
resolved